// ts-collections/index.js
"use strict";
var ArrayList_1 = require('./collection/ArrayList');
exports.ArrayList = ArrayList_1.ArrayList;
var Collection_1 = require('./collection/Collection');
exports.AbstractCollection = Collection_1.AbstractCollection;
exports.Collection = Collection_1.Collection;
var Collections_1 = require('./collection/Collections');
exports.Collections = Collections_1.Collections;
var DecoratedCollection_1 = require('./collection/DecoratedCollection');
exports.DecoratedCollection = DecoratedCollection_1.DecoratedCollection;
var Iterator_1 = require('./collection/Iterator');
exports.DefaultIterator = Iterator_1.DefaultIterator;
exports.ArrayIterator = Iterator_1.ArrayIterator;
var IteratorResultImpl_1 = require('./collection/IteratorResultImpl');
exports.IteratorResultImpl = IteratorResultImpl_1.IteratorResultImpl;
var Comparators_1 = require('./comparator/Comparators');
exports.Comparators = Comparators_1.Comparators;
var AlwaysTruePredicate_1 = require('./predicate/AlwaysTruePredicate');
exports.AlwaysTruePredicate = AlwaysTruePredicate_1.AlwaysTruePredicate;
var Predicates_1 = require('./predicate/Predicates');
exports.Predicates = Predicates_1.Predicates;
